nimare/meta/ale.py

```diff
@@ -204,59 +202,48 @@ def _compute_null_analytic(self, ma_maps):
+
+        n_zeros = np.sum(ma_values == 0, 1)
+        ma_hists[:, 0] = n_zeros
+
+        for i_exp in range(len(ma_values)):
             reduced_ma_values = ma_values[i_exp, ma_values[i_exp, :] > 0]
-            ma_hists[i_exp, 0] = n_zeros
-            ma_hists[i_exp, 1:] = np.histogram(
-                a=reduced_ma_values, bins=self.null_distributions_["histogram_bins"], density=False
-            )[0]
+            ma_hists[i_exp, 1:] = np.histogram(reduced_ma_values, bins=hist_bins, density=False)[0]
 
-        # Inverse of step size in histBins (0.0001) = 10000
-        step = 1 / np.mean(np.diff(self.null_distributions_["histogram_bins"]))
+        inv_step_size = int(np.ceil(1 / step_size))
+
+        # Normalize MA histograms to get probabilities
+        ma_hists /= ma_hists.sum(1)[:, None]
+
+        ale_hist = ma_hists[0, :].copy()
 
-        # Null distribution to convert ALE to p-values.
-        ale_hist = ma_hists[0, :]
         for i_exp in range(1, ma_hists.shape[0]):
-            temp_hist = np.copy(ale_hist)
-            ma_hist = np.copy(ma_hists[i_exp, :])
+
+            exp_hist = ma_hists[i_exp, :]
 
             # Find histogram bins with nonzero values for each histogram.
-            ale_idx = np.where(temp_hist > 0)[0]
-            exp_idx = np.where(ma_hist > 0)[0]
-
-            # Normalize histograms.
-            temp_hist /= np.sum(temp_hist)
-            ma_hist /= np.sum(ma_hist)
-
-            # Perform weighted convolution of histograms.
-            ale_hist = np.zeros(self.null_distributions_["histogram_bins"].shape[0])
-            for j_idx in exp_idx:
-                # Compute probabilities of observing each ALE value in histBins
-                # by randomly combining maps represented by maHist and aleHist.
-                # Add observed probabilities to corresponding bins in ALE
-                # histogram.
-                probabilities = ma_hist[j_idx] * temp_hist[ale_idx]
-                ale_scores = 1 - (1 - self.null_distributions_["histogram_bins"][j_idx]) * (
-                    1 - self.null_distributions_["histogram_bins"][ale_idx]
-                )
-                score_idx = np.floor(ale_scores * step).astype(int)
-                np.add.at(ale_hist, score_idx, probabilities)
+            ale_idx = np.where(ale_hist > 0)[0]
+            exp_idx = np.where(exp_hist > 0)[0]
+
+            # Compute output MA values, ale_hist indices, and probabilities
+            ale_scores = 1 - np.outer(1 - hist_bins[exp_idx], 1 - hist_bins[ale_idx]).ravel()
+            score_idx = np.floor(ale_scores * inv_step_size).astype(int)
+            probabilities = np.outer(exp_hist[exp_idx], ale_hist[ale_idx]).ravel()
+
+            # Reset histogram and set probabilities. Use at() because there can
+            # be redundant values in score_idx.
+            ale_hist = np.zeros((inv_step_size,))
+            np.add.at(ale_hist, score_idx, probabilities)
```

Comment:
Everything else looks good to me, so I assume that this is the section that is leading to the failed tests. I will try to dig into the new code, but for now I don't actually understand what's going on.

Reply:
I think the issue has to do with inconsistent rounding/casting—for the MKDA kernel, we get `score_idx` values in the top bin (i.e., 1), and they're being rounded above that. I don't think it's worth chasing this down right now, as it may go away by itself when I refactor this.